### PR TITLE
Some code improvements

### PR DESCRIPTION
**Pull request by shadowscatcher**
- modules
 - typehinting
 - requirements.txt
 - vuln descriptions are cached in internal dictionary to eliminate possible duplicate requests
 - requests are made within one [requests session](https://requests.readthedocs.io/en/master/user/advanced/#session-objects)
- markdown report generator (not used now, but implemented)

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,25 @@
+boto3==1.10.26
```
Comment (nadinelyab) [inline requirements.txt]: Hey @shadowscatcher ,
In the interest of keeping the project lightweight, could you delete any and all libraries in this file that are not required? I don't think botocore is needed, google-auth, google-api-core, and a whole bunch of others.

Comment (shadowscatcher) [reply requirements.txt]: boto3 and gcp libraries are required by aws_push.py and gcp_push.py, and botocore and google-* are their dependencies. Here's a list of dependencies installed alongside boto3, for example:

```
botocore==1.15.14
docutils==0.15.2
jmespath==0.9.5
pkg-resources==0.0.0
python-dateutil==2.8.1
s3transfer==0.3.3
six==1.14.0
urllib3==1.25.8
```

In order to completely get rid of such dependencies, we have to write corresponding upload methods from scratch without official clients.

Comment (shadowscatcher) [reply requirements.txt]: Done. Protobuf is required by boto3.

**Comment by ndisidore** [comment]
Excellent PR! Really appreciate the typing contributions as well and I think an ABC for the writer is a move in the right direction! :raised_hands: 

One thing worth mentioning: I see a few warnings regaring some of the latex start commands being interpreted as string escapes. I'd suggest double slashing these to be on the safe side.

```
❯ flake8 --select W605 contrib
contrib/report_builders/latex_report_builder.py:29:27: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:40:27: W605 invalid escape sequence '\i'
contrib/report_builders/latex_report_builder.py:40:42: W605 invalid escape sequence '\l'
contrib/report_builders/latex_report_builder.py:40:59: W605 invalid escape sequence '\l'
contrib/report_builders/latex_report_builder.py:54:38: W605 invalid escape sequence '\h'
contrib/report_builders/latex_report_builder.py:55:35: W605 invalid escape sequence '\l'
contrib/report_builders/latex_report_builder.py:56:33: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:58:38: W605 invalid escape sequence '\h'
contrib/report_builders/latex_report_builder.py:58:45: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:59:31: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:61:27: W605 invalid escape sequence '\F'
contrib/report_builders/latex_report_builder.py:65:31: W605 invalid escape sequence '\i'
contrib/report_builders/latex_report_builder.py:66:40: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:67:23: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:71:27: W605 invalid escape sequence '\i'
contrib/report_builders/latex_report_builder.py:71:42: W605 invalid escape sequence '\l'
contrib/report_builders/latex_report_builder.py:71:66: W605 invalid escape sequence '\l'
contrib/report_builders/latex_report_builder.py:74:31: W605 invalid escape sequence '\i'
contrib/report_builders/latex_report_builder.py:75:27: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:76:23: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:79:94: W605 invalid escape sequence '\l'
contrib/report_builders/latex_report_builder.py:80:23: W605 invalid escape sequence '\l'
contrib/report_builders/latex_report_builder.py:83:23: W605 invalid escape sequence '\s'
contrib/report_builders/latex_report_builder.py:86:23: W605 invalid escape sequence '\s'
contrib/report_builders/latex_report_builder.py:89:23: W605 invalid escape sequence '\s'
contrib/report_builders/latex_report_builder.py:93:23: W605 invalid escape sequence '\i'
contrib/report_builders/latex_report_builder.py:96:23: W605 invalid escape sequence '\e'
contrib/report_builders/latex_report_builder.py:97:23: W605 invalid escape sequence '\e'
```

**Comment by shadowscatcher** [comment]
@ndisidore thanks for the review and nice suggestion. Yes, the whole point of PR in the first place was to make it easier to add new report formats. HTML is ready in a separate branch of my fork, but that should be another PR.

**Comment by nadinelyab** [comment]
Looks good. Excited to see more parsers added to flan.

**Pull request merged**